### PR TITLE
Provisioner lifecycle

### PR DESCRIPTION
**Pull request by jbardin**
In verifying the lifecycle of the provisioners, it became clear that the graph nodes used to represent provisioners were of little value. The provisioner graph nodes were only responsible for initializing the provisioner (which just added it to the context's cache) and closing the provisioner, which was done inconsistently. The nodes otherwise serve no purpose, as there is no stateful configuration for a provisioner, and configuration references are all handled by the parent resource. This means we can eliminate all the provisioner nodes from the graph with no appreciable changes. To accommodate this the `EvalContext` is updated to cache provisioner instances on demand, and `CloseProvisioner` is changed to `CloseProvisioners`, which closes all open provisioners at once. 

We ensure that the builtin provisioners can be re-used by making the `Close` method a `noop`, which is OK because these are not plugins.

The mock provisioner is updated to ensure that a single instance is never used after `Close` in order to protect real plugins since we now always close the provisioners correctly. An e2e test is added to verify the builtin provisioners can be run through a full plan+apply cycle.

### CODE REVIEW

File: terraform/eval_context.go [hunk 1]
```diff
@@ -77,22 +77,17 @@ type EvalContext interface {
 	ProviderInput(addrs.AbsProviderConfig) map[string]cty.Value
 	SetProviderInput(addrs.AbsProviderConfig, map[string]cty.Value)
 
-	// InitProvisioner initializes the provisioner with the given name.
-	// It is an error to initialize the same provisioner more than once.
-	InitProvisioner(string) error
-
 	// Provisioner gets the provisioner instance with the given name (already
 	// initialized) or returns nil if the provisioner isn't initialized.
```
Comment (alisdair) [inline terraform/eval_context.go:81]: Should this comment be updated?

File: terraform/node_module_expand.go [hunk 2]
```diff
@@ -140,7 +140,9 @@ func (n *nodeExpandModule) Execute(ctx EvalContext, op walkOperation) (diags tfd
 // wait on this node.
 // Besides providing a root node for dependency ordering, nodeCloseModule also
 // cleans up state after all the module nodes have been evaluated, removing
-// empty resources and modules from the state.
+// empty resources and modules from the state, and closes a remaining
```
Comment (alisdair) [inline terraform/node_module_expand.go:143]:
```suggestion
// empty resources and modules from the state, and closes any remaining
```

**Review by alisdair — APPROVED** [review]
This looks great to me!

File: terraform/eval_context.go (hunk 1 above)
Comment (jbardin) [reply terraform/eval_context.go:81]: good catch, thanks!

**Pull request merged**